lib/net/pktdrv/pktdrv.c

```diff
@@ -342,10 +345,9 @@ enum {
 
 
 //Register access macros for XBOX
-#define	BASE	0xFEF00000
-#define REG(x)	(*((DWORD *)(BASE+(x))))
-#define REGW(x)	(*((WORD *)(BASE+(x))))
-#define REGB(x)	(*((BYTE *)(BASE+(x))))
+#define REG(x)	(*((DWORD *)((ULONG_PTR)PktdrvMmioBase+(x))))

```

Comment:
Can we create inline functions instead of these macros?

Reply:
Then we'll need to modify it in the caller as we'd have to separate reads and writes.

I think this would be an unrelated change, which should be a separate PR.

Reply:
I'm not a big fan of the way this code is written, I guess we can clean it up in a follow-up PR. I thought *this* PR was supposed to do some cleaning, but that's OK.

Another issue I see here is that we need to use the `volatile` keyword when writing to MMIO to prevent the compiler from optimizing this stuff out.

Reply:
> I'm not a big fan of the way this code is written, I guess we can clean it up in a follow-up PR. I thought this PR was supposed to do some cleaning, but that's OK.

This wasn't really a clean-up, but just a hack so xbe-loader can relocate the NIC. See https://github.com/XboxDev/nxdk/pull/100#discussion_r279218102

That's really the only purpose that I can remember.

> Another issue I see here is that we need to use the volatile keyword when writing to MMIO to prevent the compiler from optimizing this stuff out.

Yes. Should possibly happen in another PR *before* this one, just to avoid any regressions due to differences (as compiler can't assume a constant address anymore if we use a variable).

---

I've also raised some other concerns on Discord for a quick-discussion, but I'll move them here:

Moving away from macros would possibly make some operations non-atomic: `REG(NvRegDuplexMode) &= NVREG_DUPLEX_MODE_FDMASK;` which is read/modify/write would have to split this into 3 separate steps (which might affect behaviour).

Also readable code like `REG(NvRegMIIStatus)=REG(NvRegMIIStatus);` would become something like `write_reg(NvRegMIIStatus, read_reg(NvRegMIIStatus));`

I'm proposing structs for MMIO instead: `regs.NvRegMIIStatus = regs.NvRegMIIStatus;`.

However, these sort of changes should happen in a follow-up.